reference/forms/types/choice.rst

```diff
@@ -83,40 +87,98 @@ code, while the **key** is what will be shown to the user.
-        'choices' => [
+    $builder->add('category', 'choice', array(
+        'choices' => array(
             new Category('Cat1'),
             new Category('Cat2'),
             new Category('Cat3'),
             new Category('Cat4'),
-        ],
+        ),
         'choices_as_values' => true,
-        'choice_label' => function($category, $key, $index) {
-            /** @var Category $category */
+        'choice_label' => function(Category $category, $key, $value) {
             return strtoupper($category->getName());
         },
-        'choice_attr' => function($category, $key, $index) {
-            return ['class' => 'category_'.strtolower($category->getName())];
+        'choice_attr' => function(Category $category, $key, $value) {
+            return array('class' => 'category_'.strtolower($category->getName()));
         },
-        'group_by' => function($category, $key, $index) {
+        'group_by' => function(Category $category, $key, $value) {
             // randomly assign things into 2 groups
             return rand(0, 1) == 1 ? 'Group A' : 'Group B';
         },
-        'preferred_choices' => function($category, $key, $index) {
-            return $category->getName() == 'Cat2' || $category->getName() == 'Cat3';
+        'preferred_choices' => function(Category $category, $key, $value) {
+            return 'Cat2' === $category->getName() || 'Cat3' === $category->getName();
         },
-    ]);
+    ));
 
 You can also customize the `choice_name`_ and `choice_value`_ of each choice if
 you need further HTML customization.
 
+.. caution::
+
+    When dealing with objects as choices, you should be careful about how
+    string values are set to use them with the `empty_data` option.
+    In the example above, the default values are incrementing integers if the
+    ``Category`` class does not implement ``toString`` method.
+    To get a full control of the string values use the `choice_value`_ option::
+
+        $builder->add('category', 'choice', array(
+            'choices'  => array(
+            new Category('Cat1'),
+            new Category('Cat2'),
+            new Category('Cat3'),
+            new Category('Cat4'),
+            ),
+            'choices_as_values' => true,
+            'choice_value' => function(Category $category = null) {
+                if (null === $category) {
+                    return '';
+                }
+
+                return strtolower($category->getName());
+            },
+            'choice_label' => function(Category $category, $key, $value) {
+                return strtoupper($category->getName());
+            },
+            'multiple' => true,
+            'empty_data' => array('cat2'), // default submitted value
+                                           // an array because of multiple option
+        ));
+
+    Note that `choice_value`_ option set as a callable can get passed ``null``
+    when no data is preset or submitted.
+
```

Comment:
Nope, anytime no value (null) is submitted and there is no placeholder (empty string).